File: packages/api/src/models/ReadingSession.ts
/**
 * ReadingSession Model
 */

import { DataTypes, Model, Optional } from 'sequelize';
import { sequelize } from '../db';

interface ReadingSessionAttributes {
  id: string;
  userId: string;
  bookId: string;
  startedAt: Date;
  endedAt?: Date;
  duration: number;
  pagesRead: number;
  highlights: number;
  notes: number;
  isActive: boolean;
  createdAt: Date;
  updatedAt: Date;
}

interface ReadingSessionCreationAttributes extends Optional<ReadingSessionAttributes, 'id' | 'endedAt' | 'isActive' | 'createdAt' | 'updatedAt'> {}

export class ReadingSession extends Model<ReadingSessionAttributes, ReadingSessionCreationAttributes> implements ReadingSessionAttributes {
  public id!: string;
  public userId!: string;
  public bookId!: string;
  public startedAt!: Date;
  public endedAt?: Date;
  public duration!: number;
  public pagesRead!: number;
  public highlights!: number;
  public notes!: number;
  public isActive!: boolean;
  public readonly createdAt!: Date;
  public readonly updatedAt!: Date;
}

ReadingSession.init(
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    userId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: {
        model: 'users',
        key: 'id',
      },
      onDelete: 'CASCADE',
    },
    bookId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: {
        model: 'books',
        key: 'id',
      },
      onDelete: 'CASCADE',
    },
    startedAt: {
      type: DataTypes.DATE,
      allowNull: false,
      defaultValue: DataTypes.NOW,
    },
    endedAt: {
      type: DataTypes.DATE,
      allowNull: true,
    },
    duration: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 0,
    },
    pagesRead: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 0,
    },
    highlights: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 0,
    },
    notes: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 0,
    },
    isActive: {
      type: DataTypes.BOOLEAN,
      allowNull: false,
      defaultValue: false,
    },
    createdAt: {
      type: DataTypes.DATE,
      allowNull: false,
    },
    updatedAt: {
      type: DataTypes.DATE,
      allowNull: false,
    },
  },
  {
    sequelize,
    tableName: 'reading_sessions',
    timestamps: true,
    indexes: [
      { fields: ['user_id'] },
      { fields: ['book_id'] },
      { fields: ['started_at'] },
      { fields: ['user_id', 'is_active'] },
      { fields: ['user_id', 'book_id'] },
      // Covers DATE(started_at) GROUP BY queries for calendar/streak
      { fields: ['user_id', 'started_at'] },
    ],
  }
);

File: packages/server/app/models/reading_session.py
"""Reading session model."""

import uuid
from datetime import datetime
from typing import TYPE_CHECKING, Optional
from uuid import UUID

from sqlalchemy import Boolean, ForeignKey, Index, Integer, Text, text
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db import Base

if TYPE_CHECKING:
    from app.models.book import Book
    from app.models.user import User


class ReadingSession(Base):
    __tablename__ = 'reading_sessions'
    __table_args__ = (
        Index(
            'ix_reading_sessions_user_id_is_active',
            'user_id',
            'is_active',
        ),
        Index(
            'ix_reading_sessions_user_id_book_id',
            'user_id',
            'book_id',
        ),
        Index(
            'ix_reading_sessions_user_id_started_at',
            'user_id',
            'started_at',
        ),
    )

    id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
        server_default=text('gen_random_uuid()'),
    )
    user_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey('users.id', ondelete='CASCADE'),
        nullable=False,
        index=True,
    )
    book_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey('books.id', ondelete='CASCADE'),
        nullable=False,
        index=True,
    )
    started_at: Mapped[datetime] = mapped_column(
        default=datetime.utcnow,
    )
    ended_at: Mapped[Optional[datetime]] = mapped_column(
        nullable=True,
    )
    duration: Mapped[int] = mapped_column(
        Integer,
        default=0,
    )
    pages_read: Mapped[int] = mapped_column(
        Integer,
        default=0,
    )
    highlights: Mapped[int] = mapped_column(
        Integer,
        default=0,
    )
    notes: Mapped[int] = mapped_column(
        Integer,
        default=0,
    )
    summary: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True,
    )
    is_active: Mapped[bool] = mapped_column(
        Boolean,
        default=False,
    )
    created_at: Mapped[datetime] = mapped_column(
        default=datetime.utcnow,
    )
    updated_at: Mapped[datetime] = mapped_column(
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
    )

    # relationships
    user: Mapped['User'] = relationship(
        'User',
        back_populates='reading_sessions',
    )
    book: Mapped['Book'] = relationship(
        'Book',
        back_populates='reading_sessions',
    )
